client/state/data-layer/wpcom-http/utils.js

```diff
@@ -137,34 +211,43 @@ const defaultOptions = {
  * @param {Object} options configures additional dispatching behaviors
  + @param {Function} [options.fromApi] maps between API data and Calypso data
  + @param {Function} [options.onProgress] called on progress events when uploading
+ * @param {Function} [options.middleware] runs before the dispatch itself
+ * @param {Function} [options.onProgress] called on progress events when uploading
  * @returns {?*} please ignore return values, they are undefined
  */
-export const dispatchRequest = ( initiator, onSuccess, onError, options = {} ) => (
-	store,
-	action
-) => {
-	const { fromApi, onProgress } = { ...defaultOptions, ...options };
+export const dispatchRequest = ( initiator, onSuccess, onError, options = {} ) => {
+	const { fromApi, middleware, onProgress } = { ...defaultOptions, ...options };
 
-	const error = getError( action );
-	if ( undefined !== error ) {
-		return onError( store, action, error );
-	}
+	// this is an odd way of injecting middleware
```

Comment:
I don't think i follow this logic.
Why not do it the normal way and export the unmiddleware-ed version just for tests?

Reply:
Also, is there any need to have a version of `dispatchRequest` without the meta-tracking functionality? It doesn't have any weird dependencies that would need to be mocked, does it? It's only action data manipulation and adding properties to `meta.dataLayer`.

It's suspicious if something has one version for testing and another for consumption.

Reply:
some of the reason I have this is because it ended up being quite confusing when I wrote it all as a whole vs as a separate middleware. additionally, the middleware doesn't really care about things like `onSuccess` or the other handlers.

lots of it then was just keeping concerns separated. would you think it better to create this instead?

```js
export const requestDispatcher = middleware => ( initiator, … ) => { … }

export const dispatchRequest = requestDispatcher( trackRequest );
```

I could get behind that

Reply:
got behind that in 3fcea8d19